
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Error Explanations - Bluebird - W3cubDocs</title>
  
  <meta name="description" content=" You got this this error because you&#39;ve used Promise.promisify on an object, for example&#58; ">
  <meta name="keywords" content="error, explanations, bluebird">
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/bluebird/error-explanations.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-01fda2ddb8339756caccf7add5ad4cf849ab52d069bd799015c7f04f93164f64753bff0d15a49d8060b1e66e41002bb301ccadc2350937df079cea3cd52d3cca.css">
  <script src="/assets/application-d9be6f56a823612443fc15b2e027a630e02c4ad2685bb750d13fa4fae28d46c3e7f7ebb69bd4bafddf116f218f9372e9be44021d4247dc20424e2fd1ff8cef81.js" type="text/javascript"></script>
  <script src="/json/bluebird.js"></script>
  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-71174418-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>
  <script data-ad-client="ca-pub-2572770204602497" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/bluebird/" class="_nav-link" title="" style="margin-left:0;">Bluebird</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="link"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _simple">
				
				
<h1 class="post-title">Error Explanations</h1>   <ul> <li><a href="#error-promise.promisify-called-on-an-object" title="">Error: Promise.promisify called on an object</a></li> <li><a href="#error-the-promise-constructor-requires-a-resolver-function" title="">Error: the promise constructor requires a resolver function</a></li> <li><a href="#error-the-promise-constructor-cannot-be-invoked-directly" title="">Error: the promise constructor cannot be invoked directly</a></li> <li><a href="#error-expecting-an-array-a-promise-or-a-thenable" title="">Error: expecting an array, a promise or a thenable</a></li> <li><a href="#error-generatorfunction-must-be-a-function" title="">Error: generatorFunction must be a function</a></li> <li><a href="#error-fn-must-be-a-function" title="">Error: fn must be a function</a></li> <li><a href="#error-cannot-enable-long-stack-traces-after-promises-have-been-created" title="">Error: cannot enable long stack traces after promises have been created</a></li> <li><a href="#error-cannot-get-fulfillment-value-of-a-non-fulfilled-promise" title="">Error: cannot get fulfillment value of a non-fulfilled promise</a></li> <li><a href="#error-cannot-get-rejection-reason-of-a-non-rejected-promise" title="">Error: cannot get rejection reason of a non-rejected promise</a></li> <li><a href="#error-the-target-of-promisifyall-must-be-an-object-or-a-function" title="">Error: the target of promisifyAll must be an object or a function</a></li> <li><a href="#error-circular-promise-resolution-chain" title="">Error: circular promise resolution chain</a></li> <li><a href="#error-cannot-await-properties-of-a-non-object" title="">Error: cannot await properties of a non-object</a></li> <li><a href="#error-expecting-a-positive-integer" title="">Error: expecting a positive integer</a></li> <li><a href="#error-a-value-was-yielded-that-could-not-be-treated-as-a-promise" title="">Error: A value was yielded that could not be treated as a promise</a></li> <li><a href="#error-cannot-await-properties-of-a-non-object" title="">Error: cannot await properties of a non object</a></li> <li><a href="#error-cannot-promisify-an-api-that-has-normal-methods" title="">Error: Cannot promisify an API that has normal methods</a></li> <li><a href="#error-catch-filter-must-inherit-from-error-or-be-a-simple-predicate-function" title="">Error: Catch filter must inherit from Error or be a simple predicate function</a></li> <li><a href="#error-no-async-scheduler-available" title="">Error: No async scheduler available</a></li> </ul> <h2 id="error-promise.promisify-called-on-an-object">  Error: Promise.promisify called on an object </h2> <p>You got this this error because you've used <code>Promise.promisify</code> on an object, for example:</p> <pre data-language="javascript">var fs = Promise.promisify(require("fs"));
</pre> <p>Instead, use <a href="api/promise.promisifyall"><code>Promise.promisifyAll</code></a> :</p> <pre data-language="javascript">var fs = Promise.promisifyAll(require("fs"));
</pre> <h2 id="error-the-promise-constructor-requires-a-resolver-function">  Error: the promise constructor requires a resolver function </h2> <p>You got this error because you used <code>new Promise()</code> or <code>new Promise(something)</code> without passing a function as the parameter.</p> <p>If you want to wrap an API with a promise manually, the correct syntax is:</p> <pre data-language="javascript">function wrapWithPromise(parameter) {
    return new Promise(function (resolve, reject) {
        doSomethingAsync({
              error:reject,
              success:resolve
        });
    });
}
</pre> <p>Please consider reading about <a href="api/new-promise"><code>new Promise</code></a> and also consider checking out automatic <a href="api/promisification"><code>promisification</code></a> as well as <a href="api/promise.method"><code>Promise.method</code></a></p> <h2 id="error-the-promise-constructor-cannot-be-invoked-directly">  Error: the promise constructor cannot be invoked directly </h2> <p>You can get this error for several reasons:</p> <h4 id="1.-you-forgot-to-use-new-when-creating-a-new-promise-using-new-promise">  1. You forgot to use <code>new</code> when creating a new promise using <code>new Promise(resolver)</code> syntax. </h4> <p>This can happen when you tried to do something like:</p> <pre data-language="javascript">return Promise(function(resolve,reject){
       //...
})
</pre> <p>You can correct this by doing:</p> <pre data-language="javascript">return new Promise(function(resolve,reject){
       //...
})
</pre> <p>Please consider reading about <a href="api/new-promise"><code>new Promise</code></a> and also consider checking out automatic <a href="api/promisification"><code>promisification</code></a> as well as <a href="api/promise.method"><code>Promise.method</code></a></p> <h4 id="2.-you-are-trying-to-subclass-promise">  2. You are trying to subclass <code>Promise</code> </h4> <p>Bluebird does not support extending promises this way. Instead, see <a href="features#scoped-prototypes" title="">scoped prototypes</a>.</p> <h2 id="error-expecting-an-array">  Error: expecting an array, a promise or a thenable </h2> <p>The function being called expects a Promise, but is given something different. There are two main reasons why this may occur.</p> <p><strong>1. Working with collections (like arrays) but pass a single, non-collection element instead</strong></p> <p>Example:</p> <pre data-language="javascript">function returnThree(){ return 3;}

Promise.resolve(5).map(returnThree).then(function(val){
     console.log("Hello Value!",val); 
});
</pre> <p>The <code>map</code> operation is expecting an array here (or a promise on one) and instead gets the number <code>5</code>.</p> <pre data-language="javascript">function returnThree(){ return 3;}

Promise.resolve([5]).map(returnThree).then(function(val){
     console.log("Hello Value!",val); 
});
</pre> <p><code>map</code> is given an array with a single element (see <code>[5]</code> instead of <code>5</code>), so this statement will work (but is bad practice).</p>  <p><strong>2.<code>return</code> is forgotten in a 'fat' arrow / anonymous function call <code>=&gt;</code>:</strong></p> <p>When debugging or performing a one-time operation on a variable before passing it to a function, a return variable is forgotten.</p> <p>Example:</p> <pre data-language="javascript">function nextFunction(something){ return Promise.resolve(something*3); }

myFunction()
    .then(result =&gt; nextFunction(result)); // We are implicitly returning a Promise
</pre> <p>Debugging, we want to see the value of result, so we add a <code>console.log()</code> line:</p> <pre data-language="javascript">function nextFunction(something){ return Promise.resolve(something*3); }

myFunction().then(result =&gt; {
    console.log("Debug:", result);
    nextFunction(result)); // The chain is broken! We don't return anything to the .then() call
});
</pre> <p>As this is an anonymous function call, we need to <strong>return</strong> something, which is not currently happening.</p> <p>To fix, simply remember to add <code>return</code> in front of your promise-complying function:</p> <pre data-language="javascript">function nextFunction(something){ return Promise.resolve(something*3); }

myFunction().then(result =&gt; {
    console.log("Debug:", result);
    return nextFunction(result)); // The anonymous function returns the function which returns the promise .then() needs
});
</pre> <h2 id="error-generatorfunction-must-be-a-function">  Error: generatorFunction must be a function </h2> <p>You are getting this error when trying to use <a href="api/promise.coroutine"><code>Promise.coroutine</code></a> and not passing it a generator function as a parameter. For example:</p> <pre data-language="javascript">Promise.coroutine(function* () { // Note the *
    var data = yield $.get("http://www.example.com");
    var moreUrls = data.split("\n");
    var contents = [];
    for( var i = 0, len = moreUrls.length; i &lt; len; ++i ) {
        contents.push(yield $.get(moreUrls[i]));
    }
    return contents;
});
</pre> <p>Please refer to the relevant section in the documentation about <a href="api/generators"><code>Generators</code></a> in order to get usage instructions:</p> <p><strong>Note</strong>: Bluebird used to eagerly check for generators which caused problems with transpilers. Because of this, you might get an error similar to <code>TypeError: Cannot read property 'next' of undefined</code> if you pass a function instead of a generator function to Bluebird.</p> <p><a href="api/promise.coroutine"><code>Promise.coroutine</code></a> is built to work with generators to form C# like <code>async/await</code></p> <h2 id="error-fn-must-be-a-function">  Error: fn must be a function </h2> <p>You passed a non-function where a function was expected.</p> <h2 id="error-cannot-enable-long-stack-traces-after-promises-have-been-created">  Error: cannot enable long stack traces after promises have been created </h2> <p>You are getting this error because you are enabling long stack traces after a promise has already been created.</p> <p>When using <code>longStackTraces</code> the first line in your code after requiring Bluebird should be:</p> <pre data-language="javascript">Promise.config({
    longStackTraces: true
});
</pre> <p>See the API page about <a href="api/promise.longstacktraces"><code>Promise.longStackTraces</code></a></p> <h2 id="error-cannot-get-fulfillment-value-of-a-non-fulfilled-promise">  Error: cannot get fulfillment value of a non-fulfilled promise </h2> <p>You can get this error when you're trying to call <code>.value</code> or <code>.error</code> when inspecting a promise where the promise has not been fulfilled or rejected yet.</p> <p>For example:</p> <pre data-language="javascript">var p = Promise.delay(1000);
p.inspect().value();
</pre> <p>Consider using <a href="api/ispending"><code>.isPending()</code></a> <a href="api/isfulfilled"><code>.isFulfilled()</code></a> and <a href="api/isrejected"><code>.isRejected()</code></a> in order to inspect the promise for status.</p> <p>Please consider reading more about <a href="api/synchronous-inspection"><code>synchronous inspection</code></a></p> <h2 id="error-cannot-get-rejection-reason-of-a-non-rejected-promise">  Error: cannot get rejection reason of a non-rejected promise </h2> <p>You can get this error when you're trying to call <code>.value</code> or <code>.error</code> when inspecting a promise where the promise has not been fulfilled or rejected yet.</p> <p>For example:</p> <pre data-language="javascript">var p = Promise.delay(1000);
p.inspect().value();
</pre> <p>Consider using <a href="api/ispending"><code>.isPending()</code></a> <a href="api/isfulfilled"><code>.isFulfilled()</code></a> and <a href="api/isrejected"><code>.isRejected()</code></a> in order to inspect the promise for status.</p> <p>Please consider reading more about <a href="api/synchronous-inspection"><code>synchronous inspection</code></a></p> <h2 id="error-the-target-of-promisifyall-must-be-an-object-or-a-function">  Error: the target of promisifyAll must be an object or a function </h2> <p>This can happen when you are calling <a href="api/promise.promisifyall"><code>Promise.promisifyAll</code></a> on a function and invoking it instead of passing it.</p> <p>In general, the usage of <a href="api/promise.promisifyall"><code>Promise.promisifyAll</code></a> is along the lines of <code>var fs = Promise.promisifyAll(require("fs"))</code>.</p> <p>Consider reading the section about <a href="api/promisification"><code>promisification</code></a></p> <h2 id="error-circular-promise-resolution-chain">  Error: circular promise resolution chain </h2> <p>This usually happens when you have a promise that resolves or rejects with itself.</p> <p>For example: <code>var p = Promise.delay(100).then(function(){ return p});</code> .</p> <p>In this case, the promise resolves with itself which was is not intended.</p> <p>This also happens when implementing live-updating models with a <code>.then</code> method that indicates when the model is "ready". A promise is a process, it starts and it ends.</p> <p>Promises do not aim to solve such live updating problems directly. One option would be to use an intermediate promise - for example a <code>.loaded</code> property on the model that fulfills with nothing.</p> <p>resolving it with itself tells it "it is done when it is done"</p> <h2 id="error-cannot-await-properties-of-a-non-object">  Error: cannot await properties of a non-object </h2> <p>The <code>.props</code> method expects to receive an object.</p> <p>For example:</p> <pre data-language="javascript">Promise.props({
    pictures: getPictures(),
    comments: getComments(),
    tweets: getTweets()
}).then(function(result){
    console.log(result.tweets, result.pictures, result.comments);
});
</pre> <p>This happens when a non object value or a promise that resolves with something that is not an object is being passed instead.</p> <h2 id="error-expecting-a-positive-integer">  Error: expecting a positive integer </h2> <p>This happens when you call <code>.some</code> passing it a negative value or a non-integer.</p> <p>One possible cause is using <code>.indexOf</code> which returns <code>-1</code> when it doesn't find the value being searched for.</p> <p>Please consider reading the API docs for <a href="api/some"><code>.some</code></a></p> <h2 id="error-a-value-was-yielded-that-could-not-be-treated-as-a-promise">  Error: A value was yielded that could not be treated as a promise </h2> <p>You are getting this error because you have tried to <code>yield</code> something in a coroutine without a yield handler, for example:</p> <pre data-language="javascript">var coroutine = Promise.coroutine(function*(){
    var bar = yield "Foo";
    console.log(bar);
});
</pre> <p>The solution is to either convert it to a promise by calling <code>Promise.resolve</code> on it or <code>Promise.promisify</code> if it's a callback:</p> <pre data-language="javascript">var coroutine = Promise.coroutine(function*(){
    var bar = yield Promise.resolve("Foo");
    console.log(bar);
});
</pre> <p>Or to use <a href="api/promise.coroutine.addyieldhandler"><code>Promise.coroutine.addYieldHandler`</code></a> to teach <a href="api/promise.coroutine"><code>Promise.coroutine</code></a> to accept these sort of values.</p> <h2 id="error-cannot-await-properties-of-a-non-object">  Error: cannot await properties of a non object </h2> <p>The <code>.props</code> method expects to receive an object.</p> <p>For example:</p> <pre data-language="javascript">Promise.props({
    pictures: getPictures(),
    comments: getComments(),
    tweets: getTweets()
}).then(function(result){
    console.log(result.tweets, result.pictures, result.comments);
});
</pre> <p>This happens when a non object value or a promise that resolves with something that is not an object is being passed instead.</p> <h2 id="error-cannot-promisify-an-api-that-has-normal-methods">  Error: Cannot promisify an API that has normal methods </h2> <p>This error indicates you have tried to call <a href="api/promise.promisifyall"><code>Promise.promisifyAll</code></a> on an object that already has a property with the <code>Async</code> suffix:</p> <pre data-language="javascript">var myApi = { foo: function(cb){ ... }, fooAsync(cb) { ... }
</pre> <p>This is because Bluebird adds the <code>Async</code> suffix to distinguish the original method from the promisified one, so <code>fooAsync</code> would have been overridden. In order to avoid this - either rename <code>fooAsync</code> before promisifying the API, or call <a href="api/promise.promisify"><code>Promise.promisify</code></a> manually on select properties.</p> <p>You may also use the custom suffix option to choose another suffix that doesn't result in conflicts.</p> <p>If you find this issue in a common library please <a href="https://github.com/petkaantonov/bluebird/issues/new" title="">open an issue</a>.</p> <h2 id="error-catch-filter-must-inherit-from-error-or-be-a-simple-predicate-function">  Error: Catch filter must inherit from Error or be a simple predicate function </h2> <p>Bluebird supports typed and predicate <a href="api/catch"><code>.catch()</code></a> calls]. However in order to use the typed/predicate catch syntax for error handling you must do one of two things.</p> <p>Pass it a constructor that inherits from <code>Error</code>:</p> <pre data-language="javascript">}).catch(ReferenceError, function(e) { // this is fine
}).catch(Array, function(e) { // arrays don't capture stack traces
</pre> <p>This is to enable better stack trace support and to have more consistent and logical code.</p> <p>Alternatively, if you provide it a predicate be sure it's a simple function:</p> <pre data-language="javascript">}).catch(function(e){ return false; }, function(e) { // this catches nothing
}).catch(function(e){ return e.someProp = 5; }, function(e) { // this is fine
</pre> <p>Please see the API docs of <a href="api/catch"><code>.catch()</code></a> on how to use predicate catches.</p> <h2 id="error-no-async-scheduler-available">  Error: No async scheduler available </h2> <p>Async scheduler is a function that takes a callback function and calls the callback function as soon as possible, but asynchronously. For example <code>setTimeout</code>.</p> <p>By default bluebird only tries a few common async schedulers, such as <code>setTimeout</code>, <code>process.nextTick</code> and <code>MutationObserver</code>. However if your JavaScript runtime environment doesn't expose any of these, you will see this error.</p> <p>You may use <a href="api/promise.setscheduler"><code>Promise.setScheduler</code></a> to pass a custom scheduler that your environment supports. For example in DukTape:</p> <pre data-language="javascript">Promise.setScheduler(function(fn){ // fn is what to execute
    var timer = uv.new_timer.call({});
    uv.timer_start(timer, 0, 0, fn); // add the function as a callback to the timer
});
</pre>
<div class="_attribution">
  <p class="_attribution-p">
    © 2013–2018 Petka Antonov<br>Licensed under the MIT License.<br>
    <a href="http://bluebirdjs.com/docs/error-explanations.html" class="_attribution-link">http://bluebirdjs.com/docs/error-explanations.html</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
